### PR TITLE
Anpo refcount now uses interlocked

### PR DESCRIPTION
**Pull request by anporumb**

### CODE REVIEW

File: interfaces/inc/azure_c_pal/refcount.h [hunk 1]
```diff
@@ -26,7 +26,7 @@ will interact with deallocated memory / resources resulting in an undefined beha
 #include "azure_c_pal/gballoc.h"
 
 // Include the platform-specific file that defines atomic functionality
-#include "refcount_os.h"
+#include "azure_c_pal/interlocked.h"
```
Comment (mattdurak) [inline interfaces/inc/azure_c_pal/refcount.h:29]: yay for switching this over! Can we move this header (refcount.h) to the "common" folder for "things that we need in this repo as dependencies but aren't really a PAL"? I think the "interfaces" folder should just be for PAL interfaces that have different OS implementations

File: interfaces/inc/azure_c_pal/refcount.h [hunk 2]
```diff
@@ -108,22 +108,9 @@ then in order to see "count" / "counted" this is a good string to paste in Visua
...
-#ifndef INIT_REF_VAR
-#error refcount_os.h does not define INIT_REF_VAR
-#endif // !INIT_REF
-
-#define INC_REF(type, var) INC_REF_VAR(((REFCOUNT_TYPE(type)*)((unsigned char*)var - offsetof(REFCOUNT_TYPE(type), counted)))->count)
-#define DEC_REF(type, var) DEC_REF_VAR(((REFCOUNT_TYPE(type)*)((unsigned char*)var - offsetof(REFCOUNT_TYPE(type), counted)))->count)
-#define INIT_REF(type, var) INIT_REF_VAR(((REFCOUNT_TYPE(type)*)((unsigned char*)var - offsetof(REFCOUNT_TYPE(type), counted)))->count)
+#define INC_REF(type, var) interlocked_increment(&((REFCOUNT_TYPE(type)*)((unsigned char*)var - offsetof(REFCOUNT_TYPE(type), counted)))->count)
+#define DEC_REF(type, var) interlocked_decrement(&((REFCOUNT_TYPE(type)*)((unsigned char*)var - offsetof(REFCOUNT_TYPE(type), counted)))->count)
+#define INIT_REF(type, var) do { ((REFCOUNT_TYPE(type)*)((unsigned char*)var - offsetof(REFCOUNT_TYPE(type), counted)))->count  = 1; }while(0)
```
Comment (mattdurak) [inline interfaces/inc/azure_c_pal/refcount.h:113]: >->count  = 1; [](start = 119, length = 13)

interlocked_exchange

File: linux/inc/refcount_os.h [hunk 3]
```diff
@@ -1,79 +0,0 @@
-// Copyright (c) Microsoft. All rights reserved.
-// Licensed under the MIT license. See LICENSE file in the project root for full license information.
-
-// This file gets included into refcount.h as a means of extending the behavior of
-// atomic increment, decrement, and test.
-#ifndef REFCOUNT_OS_H__LINUX
```
Comment (mattdurak) [inline linux/inc/refcount_os.h:6]: >REFCOUNT_OS_H__LINUX [](start = 8, length = 20)

cleanup :)

File: linux/tests/reals_linux/real_interlocked.h [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+#else
+#include <stdint.h>
+#include <stddef.h>
+#endif
+
+#include "azure_macro_utils/macro_utils.h"
+
+#define R2(X) REGISTER_GLOBAL_MOCK_HOOK(X, real_##X);
+
+#ifdef _WIN64
```
Comment (mattdurak) [inline linux/tests/reals_linux/real_interlocked.h:18]: I don't think we need this for Linux...

File: linux/tests/reals_linux/real_interlocked_renames.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+#define interlocked_add                        real_interlocked_add
+#define interlocked_add_64                     real_interlocked_add_64
+#define interlocked_and                        real_interlocked_and
+#define interlocked_and_16                     real_interlocked_and_16
+#define interlocked_and_64                     real_interlocked_and_64
+#define interlocked_and_8                      real_interlocked_and_8
+#define interlocked_compare_exchange           real_interlocked_compare_exchange
+#ifdef _WIN64                          
+#define interlocked_compare_exchange_128       real_interlocked_compare_exchange_128
+#endif
```
Comment (mattdurak) [inline linux/tests/reals_linux/real_interlocked_renames.h:12]: should also remove

File: win32/tests/reals_win32/real_interlocked.h [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+#include <stddef.h>
+#endif
+
+#include "azure_macro_utils/macro_utils.h"
+
+#define R2(X) REGISTER_GLOBAL_MOCK_HOOK(X, real_##X);
+
+#ifdef _WIN64 
+#define REGISTER_INTERLOCKED_GLOBAL_MOCK_HOOK()     \
+    MU_FOR_EACH_1(R2,                               \
```
Comment (mattdurak) [inline win32/tests/reals_win32/real_interlocked.h:20]: makes me wonder if we could have something simpler with less repeated lines:

define REGISTER_BASE (has everything except 128)

if win64
define REGISTER_INTERLOCKED... REGISTER_BASE and 128
else
define REGISTER_INTERLOCKED... REGISTER_BASE 
endif

File: interfaces/inc/azure_c_pal/refcount.h (hunk 1 above)
Comment (anporumb) [reply interfaces/inc/azure_c_pal/refcount.h:29]: Alright, moved it

---
In reply to: [460141738](https://github.com/Azure/azure-c-pal/pull/16#discussion_r460141738) [](ancestors = 460141738)

File: interfaces/inc/azure_c_pal/refcount.h (hunk 2 above)
Comment (anporumb) [reply interfaces/inc/azure_c_pal/refcount.h]: Absolutely, clearly, totally, completely 100% agree.

---
In reply to: [460142297](https://github.com/Azure/azure-c-pal/pull/16#discussion_r460142297) [](ancestors = 460142297)

File: linux/tests/reals_linux/real_interlocked.h (hunk 4 above)
Comment (anporumb) [reply linux/tests/reals_linux/real_interlocked.h]: "copy&paste is a design mistake" [Doc - approx 2007]

fixed

---
In reply to: [460143077](https://github.com/Azure/azure-c-pal/pull/16#discussion_r460143077) [](ancestors = 460143077)

File: linux/tests/reals_linux/real_interlocked_renames.h (hunk 5 above)
Comment (anporumb) [reply linux/tests/reals_linux/real_interlocked_renames.h]: removed

---
In reply to: [460143178](https://github.com/Azure/azure-c-pal/pull/16#discussion_r460143178) [](ancestors = 460143178)

File: win32/tests/reals_win32/real_interlocked.h (hunk 6 above)
Comment (anporumb) [reply win32/tests/reals_win32/real_interlocked.h]: Yup, clearly works :) Thanks!

---
In reply to: [460144666](https://github.com/Azure/azure-c-pal/pull/16#discussion_r460144666) [](ancestors = 460144666)

**Pull request merged**